### PR TITLE
Microos snapshot stage 2

### PR DESCRIPTION
**Pull request by ifeulner**
Hi, I don't know why but the last pull request has vanished..

Anyway, this know removes the initial snapshot and just uses the packer based snapshot for all created servers now, including the cluster autoscaler nodes. 
Please have a look and test it, hadn't hat much time for it yet. 

But as recently came the request, to make this work also in terraform cloud, I see the following alternatives:

1.  fall back to have the `packer` step outside of the terraform script (initial idea). Needs some additional "glue" scripting
2. avoid packer, do the snapshot with terraform - either by a separate script or include it here. Needs some additional "glue" scripting
3. Include the snapshot creation in the "host" module. Should be possible (when no `image_id`given, create first a snapshot, otherwise use the image_id. But Increases the complexity of the script

For me, using packer/ or a separate terraform script  to create the snapshot and assume the image_id in the terraform script would be the simpliest, cleanest way, as we can get rid of a lot of code in the main script. 

What do you think?

### CODE REVIEW

File: templates/autoscaler-cloudinit.yaml.tpl [hunk 1]
```diff
@@ -121,5 +116,12 @@ runcmd:
...
-- [systemctl, 'start', 'k3s-agent']
+# enable k3s-agent service
+- [systemctl, 'enable', 'k3s-agent']
+
+# reboot!
+power_state:
+    delay: now
+    mode: reboot
+    message: MicroOS rebooting to reflect changes
+    condition: true
```
Comment (mysticaltech) [inline templates/autoscaler-cloudinit.yaml.tpl:127]: @ifeulner Why is this needed? As it already boots into the correct btrfs snapshot.

File: packer.tf [hunk 2]
```diff
@@ -0,0 +1,44 @@
+locals {
+  packages = length(local.packages_to_install) > 0 ? "-var packages_to_install=${local.packages_to_install}" : ""
+}
+data "hcloud_image" "microos_image" {
+  with_selector = "microos-snapshot=yes,creator_id=${null_resource.packer.id}"
+  most_recent   = true
+}
+
+resource "null_resource" "packer" {
```
Comment (mysticaltech) [inline packer.tf:9]: Maybe the backend terraform mechanism can happen here?!

File: templates/autoscaler-cloudinit.yaml.tpl (hunk 1 above)
Comment (ifeulner) [reply templates/autoscaler-cloudinit.yaml.tpl:127]: But within the first boot the install of k3s happens, which will includes a new transactional-update. That's why also the k3s service is only enabled and not started.

File: agents.tf [hunk 3]
```diff
@@ -8,6 +8,7 @@ module "agents" {
   for_each = local.agent_nodes
 
   name                         = "${var.use_cluster_name_in_node_name ? "${var.cluster_name}-" : ""}${each.value.nodepool_name}"
+  microos_image_id             = data.hcloud_image.microos_image.id
```
Comment (mysticaltech) [inline agents.tf:11]: Since at other places snapshot_id is used, we should keep the naming consistent, easy fix.

File: templates/autoscaler-cloudinit.yaml.tpl (hunk 1 above)
Comment (mysticaltech) [reply templates/autoscaler-cloudinit.yaml.tpl:127]: @ifeulner Yes, but what I do not understand is why the reboot at the end of cloud-init since it's not even booted yet.

Comment (mysticaltech) [reply templates/autoscaler-cloudinit.yaml.tpl:127]: Our install of k3s does not do a transactional-update.

**Comment by mysticaltech** [comment]
@ifeulner About packer and terraform, I'm all for 1. and it would be a pleasure working on a small bash script to make the magic happen.

So we would need pure terraform logic that would kick in if not snapshot is detected, meaning it did not run via the init.sh script that calls packer.

This is great, because we can do plenty of things like ensure that all the required CLIs are present, like of course terraform, packer and ideally also hcloud.

And ideally, we would all it via curl to always get the latest and greatest, like what k3s does. Honesly, bash is super easy to write with ChatGPT, like it ridiculous. So that's an ideal use case!

Also I really like the fact that it's optional. Folks that want to use packer then terraform directly can! And they can even use terraform directly and bypass packer to fetch the image online like we do now.

So really this bash script is a layer just for adding UX goodies, like in the future automatically fetching mirrors and testing them even. If we push the envelope we could even install homebrew for them and install all the required CLIs if not present. 

But at first let's keep it simple! Just save them to type the both the packer and terraform command is a great start. And called in the form of:

```
curl -sfL https://kube-hetzner.github.io/install | sh -
```

**Comment by mysticaltech** [comment]
About the previous PR, it's here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/557 and you yourself closed it.

And thanks for this new one, looking great! Will test ASAP 🙏 🚀

**Comment by mysticaltech** [comment]
@ifeulner Let's forget the bash proposal above for now and instead use a `snapshot_deploy_method` variable, that would default to packer integrated and called via local-exec as you do already, but also have a pure terraform version like in the current master.

File: modules/host/main.tf [hunk 4]
```diff
@@ -60,33 +59,6 @@ resource "hcloud_server" "server" {
     EOT
   }
 
-  # Install MicroOS
```
Comment (mysticaltech) [inline modules/host/main.tf:63]: @ifeulner So this logic could come back and execute based on the value 
of `snapshot_deploy_method` if it's equal to "terraform" for instance.

**Comment by mysticaltech** [comment]
@ifeulner Please let me know what you think of the above proposals. Also, to keep things well compartimentalized, I propose moving the packer logic to the host module, or a new module of it's own, along with the terraform snapshot creation logic. Basically, making sure that each concerns is well compartimentalized, while staying compatible with terraform cloud when the right options are chosen (in our case `snapshot_deploy_method="terraform"` for instance).

**Comment by mysticaltech** [comment]
One last thing, and maybe we can clarify this in a subsequent PR is that, we need to clarify which variables can be changed after the initial apply, with or without `terraform destroy -target null_ressource.kustomization`

**Comment by ifeulner** [comment]
> @ifeulner Please let me know what you think of the above proposals. Also, to keep things well compartimentalized, I propose moving the packer logic to the host module, or a new module of its own, along with the terraform snapshot creation logic.
> 
> Basically, making sure that each concerns is well compartmentalized, while staying compatible with terraform cloud when the right options are chosen (in our case `snapshot_deploy_method="terraform"` for instance).

Hi @mysticaltech there are multiple options now, but it‘s more about „packaging“ and not functionality anymore.
I would prefer to keep it as simple as possible, though. So Option 1 would be my preference. 
But I understand you want to keep the terraform-only way, right?

**Comment by mysticaltech** [comment]
@ifeulner Yes, option one is fine, but let definitely with support for terraform only! 🙏

**Comment by otavio** [comment]
Using external logic for the snapshot allows for decoupling and easier code maintenance, but more critical is simplifying the snapshot update without needing to recreate the cluster.

**Comment by mysticaltech** [comment]
@ifeulner I created this new branch [snapshot-deploy](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/tree/snapshot-deploy) and opened #598, we can continue there and more people can participate if they want to.

@gthieleb Thanks for sharing your experience, I like your ideas, reducing the remote-execs if possible would be ideal, please don't hesitate to open a PR targeting the new branch mentioned above.

@aleksasiriski Same for you man, if inspired the new branch above should be the target.

Let's make this awesome folks! 🙏
